### ISSUE TITLE
Html 3.2 "type" attribute warning incorrect 

### ISSUE THREAD

**Issue by dellgreen** [Obsidian-StudiosInc/jtidy#1]
When checking Html 3.2 document, it logs out warnings that a "type" attribute is missing which is incorrect for the version it is checking.

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">

Error logged is: "line 1 column 167 - Warning: <link> lacks "type" attribute"

In html 3.2 the type attribute on "link" elements and possibly other elements are not valid

**Comment by wltjr** [comment]
That is likely a bug resulting from the change another made to support HTML 5. I forked a couple forks. I needed a tagged version, and none had tags. I guess I will take over maintaining this. It wasn't my intention but who knows. Anyway thanks for the report. I may take a look at fixing this at some point.

**Comment by dellgreen** [comment]
Cool thanks, yer the problem exists in the original jtidy938 from source forge which lead me to your fork which seemed more up to date and the one to follow :) If you have a rough idea which class/s i would need to look at for html 3.2 checks i could take a crack at fixing it and submit a pull request?

**Comment by wltjr** [comment]
Sure logical path. This may become the supported version via community effort. I have tried to get some original authors I found on Github associated with other forks unsuccessfully. No clue if the authors of this are on Github or not.

I have no clue as to the code base. If I could give you a suggestion on that front I surely would. But you likely know more about the code based than I do. Feel free to take a shot at it if you like. Otherwise I will take a look time permitting. I will likely setup Travis and update Readme, to make it more official of a project/fork.

**Comment by dellgreen** [comment]
ok, many thanks :)

**Comment by wltjr** [comment]
Sure thing! I think I found one of the original authors @aditsu. I sent them an invite to collaborate.

**Comment by wltjr** [comment]
Taking a quick look, it seems that Mappings are in the [AttributeTable](https://github.com/Obsidian-StudiosInc/jtidy/blob/master/src/main/java/org/w3c/tidy/AttributeTable.java#L308). Which seems to be set to 3.2.
```java
new Attribute("type", Dict.VERS_FROM32, AttrCheckImpl.TYPE)
```

Not sure if we can change Dict.VERS_FROM32 to something else to fix the issue. Or if that does not address the issue. Then maybe need to implement a check for Type, its presently set to null in [AttrCheckImp](https://github.com/Obsidian-StudiosInc/jtidy/blob/master/src/main/java/org/w3c/tidy/AttrCheckImpl.java#L177). May need to code up a check method, CheckType(), and change that from null to that method.

That should give you some stuff to play with to test out a fix. I am not familiar with usage, and need a 3.2 file to test against etc.

**Comment by dellgreen** [comment]
ok great, thanks for the heads up

**Comment by dellgreen** [comment]
Do all the tests pass for you, as just running maven test goal, produces over 200 unit test failures for me?

one assumes the fix would be changing 

`new Attribute("type", Dict.VERS_FROM32, AttrCheckImpl.TYPE)
`
to

new Attribute("type", Dict.VERS_HTML40, AttrCheckImpl.TYPE)

as according to the html specs if I am reading correctly, the type attribute didn't get introduced until version 4

**Comment by wltjr** [comment]
That is what I was thinking. I have not run the tests, I tend to avoid such for my needs. I needed this as a dependency for other package, [JMeter](https://github.com/Obsidian-StudiosInc/os-xtoo/blob/master/dev-java/jmeter-core/jmeter-core-9999.ebuild#L38). Which I have not finished packaging the rest of JMeter. How I build [JTidy](https://github.com/Obsidian-StudiosInc/os-xtoo/blob/master/dev-java/jtidy/jtidy-9999.ebuild), JMeter, and the rest is all Gentoo specific, not using Maven, Ant, or Gradle. Just straight javac with bash wrappers, that take it from ebuild to command. Also portage that handles dependency resolution and invokes all the bash stuff.

The tests maybe outdated and need to be updated, no clue there. I would rely more on testing what ever was giving you that error before to see if that resolves it. I tend to rely more on real usage than tests. But I am packaging, not developing. I do consider tests beneficial and useful for development. Though take their own time to develop, maintain, etc.

The whole package likely needs love. I just needed a tagged version, saw another had made some changes, so grabbed their version for a newer tag. Seems its not been maintained in some time. No clue as to the state of tests when the project was last active. They may have been failing then.

**Comment by wltjr** [comment]
I added and activated [Travis](https://travis-ci.org/Obsidian-StudiosInc/jtidy) and [Shippable](https://app.shippable.com/github/Obsidian-StudiosInc/jtidy/dashboard) CI. It build fine under Oracle JDK 8, but failed under 9. Seems the tests passed under 8. I removed source/target from both build.xml and pom.xml files. Seems to be the issue under 9.

**Comment by dellgreen** [comment]
hmm, i have tried running tests on java 7,8 runtime and they all fail. I have found the problem.
It appears that the .msg files that are used to assert the output of a validation contains an extra message in relation to version jtidy version info, however the current version of JTidy no longer produces this message meaning that the generated output and that it asserts against no longer matches.
The fix is in all .msg files to remove the message entry that contains the version info (typically message 1). When i get time i'll fix them up, as there are over 200 files to fix, so will take a little time.

I have had a look at the travis build, and combing thru the logs, the tests are failing its just that the build is not configured to fail on failing tests. :)

**Comment by wltjr** [comment]
Yes the tests are failing, I did not catch that at the end I was looking further up on one run. Anyway I updated some of the build and what not. We could always look to set that version. Pretty sure we just need to [update 1 properties file](https://github.com/Obsidian-StudiosInc/jtidy/blob/master/src/main/resources/jtidy.properties). That is what has the version. I would assume maven to set that on build.

**Comment by wltjr** [comment]
I can always add you to the project for direct commits and branching if you like. It is not my code. I have no more rights to it than you or anyone else. I assume someday to move it to one of the authors or like github.com/jtidy org. Not really familiar with this package, just needed it for JMeter. Which I haven't even got fully packaged to use. I am not really using it at all...

**Comment by dellgreen** [comment]
oh ok, if that would work better for you. Currently i have forked yours to try and fix the issues for the time being

**Comment by wltjr** [comment]
I am not sure I have interest in the library for direct usage. It is just a dependency. If you have interest I would rather not stand in your way or others to further and fix issues. Not that I am in the way, just not really helping so much to fix the issues. Which you are working on fixing. Thus I figure just give you direct ability to fix and further. Like I sent invites to others, rather see the project restored and come to life. If others still use it. Which seems its still used by JMeter so likely many still using it.

I was thinking maybe create a jitdy organization and transfer the repo there. Would be like github.com/jtidy/jtidy. Then giving others access, past developers, others who forked on github. Make it a community maintained project basically. Which you and I would be part of said community.

**Comment by dellgreen** [comment]
sounds good to me. many thanks :)

**Comment by wltjr** [comment]
Sure thing, may post on user list. Seems some were asking about it years ago, wanting new release on maven, and other stuff. All good reasons to revive the project. I sent out invites to all I could find on Github from original project on S.F.  No clue if they will join, boot us, etc. I granted all admin. I prefer less restrictions amongst project maintainers. If it becomes an issue that can change, but I doubt it will be one.

Next will be setting up Travis, Shippable, and I guess I will do SonarQube as well. Not sure about Coverity, not to happy with them at the moment. Plus their scan site has been down for most of the week. The new organization is not showing in Github or Shippable yet. Once it does I will update Travis to use that stuff instead.

**Comment by wltjr** [comment]
I also invited those I forked from. Just to reunite all, rather a team effort than any one on their own.

**Comment by wltjr** [comment]
Got everything all setup, Travis and Shippable are switched over and got SonarQube scan going, shields/links on readme. Lots of stuff to address for the scan but not all that is crucial, just house keeping. I sent invites to main organization. If you want it to show up under your github profile. I also sent that to all past authors/admins, but not forked repo owners. They do have admin access to repo just not organization.

Anyway that concludes my part for now. Still here, and will help out. But rest of the stuff is up to you. At least on this issue and fixing tests. I opened another for License, and maybe need one for website. Do a github hosted site vs s.f. Though not sure about lists, etc. At some point once ready can see about hooking into maven and gradle repos for updated jars for others. All in time.

**Comment by dellgreen** [comment]
Excellent, great work!

**Comment by wltjr** [comment]
Thanks, its all pretty minor. I tend to like doing the CI stuff, and rather see the stuff be used than not. No costs other than time to setup, and pretty worth while in the long run. Once closer to a release I will update stuff so upon tag it will auto deploy to maven/gradle and put a jar here under releases. I really like that aspect of automation. Makes doing a release as difficult as a git tag.

**Comment by wltjr** [comment]
Feel free to change the code as needed. Just try not to break API compatibility for other stuff, as I use this as a dependency, Rather not have to fix other stuff per changes here. But change things as you see fit. My only request is to keep any commits as tiny as possible limited to the least amount of files etc. I care less if there are hundreds. Easier to follow than giant changes.

No real need to work in  branch or other, but that is up to you on your work flow. I will not be coding much here, so should not have any issues there. Not till others join and start seeking to modify code. No requirements, but welcome to take over to lead the project as you see fit till another comes along who wants the job.

Thanks again for working on fixes and getting this back into good shape at least tests wise!
